### PR TITLE
New hypre ILU solver implementation and several updates to MGR solver

### PR DESCRIPTION
**Pull request by oseikuffuor1**

### CODE REVIEW

**Comment by liruipeng** [comment]
@oseikuffuor1 There were errors in the GPU regression test. Would you like to see them together? Thanks!

**Comment by oseikuffuor1** [comment]
@liruipeng yes, I will stop by your office in a bit, so we can go through them. Thanks

**Review by rfalgout — APPROVED** [review]
This all looks good to me.  Thanks!

**Comment by oseikuffuor1** [comment]
Hi all, 
This PR is almost ready to be merged. It passes the GPU tests and pending completion of the machine_tux test. If you have not approved this yet, please take a quick look and approve it if you have no concerns. It includes all of @victorapm 's changes. I would like to merge this to master this week. The changes outside MGR and ILU should be minimal.  Thanks.

**Review by ulrikeyang — APPROVED** [review]
I glanced through it, but did not review in detail.

**Review by victorapm — APPROVED** [review]
Hi @oseikuffuor1,

I looked at the files that I touched in my previous PR and they look fine for me.

**Review by liruipeng — APPROVED** [review]

File: src/parcsr_ls/_hypre_parcsr_ls.h [hunk 1]
```diff
@@ -1790,20 +1802,137 @@ HYPRE_Int hypre_MGRSetMaxCoarseLevels( void *mgr_vdata, HYPRE_Int maxlev );
...
+HYPRE_Int hypre_ILUMinHeapAddIRIi(HYPRE_Int *heap, HYPRE_Real *I1, HYPRE_Int *Ii1, HYPRE_Int len);
+HYPRE_Int hypre_ILUMaxHeapAddRabsIIi(HYPRE_Real *heap, HYPRE_Int *I1, HYPRE_Int *Ii1, HYPRE_Int len);
+HYPRE_Int hypre_ILUMaxrHeapAddRabsI(HYPRE_Real *heap, HYPRE_Int *I1, HYPRE_Int len);
+HYPRE_Int hypre_ILUMinHeapRemoveI(HYPRE_Int *heap, HYPRE_Int len);
+HYPRE_Int hypre_ILUMinHeapRemoveIIIi(HYPRE_Int *heap, HYPRE_Int *I1, HYPRE_Int *Ii1, HYPRE_Int len);
+HYPRE_Int hypre_ILUMinHeapRemoveIRIi(HYPRE_Int *heap, HYPRE_Real *I1, HYPRE_Int *Ii1, HYPRE_Int len);
+HYPRE_Int hypre_ILUMaxHeapRemoveRabsIIi(HYPRE_Real *heap, HYPRE_Int *I1, HYPRE_Int *Ii1, HYPRE_Int len);
+HYPRE_Int hypre_ILUMaxrHeapRemoveRabsI(HYPRE_Real *heap, HYPRE_Int *I1, HYPRE_Int len);
+HYPRE_Int hypre_ILUMaxQSplitRI(HYPRE_Real *array, HYPRE_Int *I, HYPRE_Int left, HYPRE_Int bound, HYPRE_Int right);
+HYPRE_Int hypre_ILUMaxQSplitRabsI(HYPRE_Real *array, HYPRE_Int *I, HYPRE_Int left, HYPRE_Int bound, HYPRE_Int right);
```
Comment (jedbrown) [inline src/parcsr_ls/_hypre_parcsr_ls.h:1869]: These break any (transitive) include in compilation units that have `<complex.h>`. Need to rename `I` to `I1` as in the other prototypes.

Comment (liruipeng) [reply src/parcsr_ls/_hypre_parcsr_ls.h:1869]: > These break any (transitive) include in compilation units that have `<complex.h>`. Need to rename `I` to `I1` as in the other prototypes.

Good point. Thanks!

Comment (jedbrown) [reply src/parcsr_ls/_hypre_parcsr_ls.h:1869]: See my PR #101 , which gets the spack petsc@develop hypre@develop build going.

**Pull request merged**